sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -267,6 +268,23 @@ case class FileSourceScanExec(
     }
   }
 
+  /**
+   * A bucket can be coalesced if the number of buckets for this relation is divisible
+   * by the number of buckets on the other side of table for join.
+   */
+  private lazy val coalescedNumBuckets: Option[Int] = {
```

Comment:
Could we compute this value in the optimizer rule side and then set this value in a constructor of `DataSourceScanExec` at that time? Probably, I think we don't need to inject the value below in relation options.